src/ImageSharp.Drawing/Processing/PathGradientBrush.cs

```diff
@@ -0,0 +1,219 @@
+                .Select(s => new Path(s))
+                .Select((path, i) => new Edge(path, ColorAt(i), ColorAt(i + 1)))
+                .ToList();
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="PathGradientBrush"/> class.
+        /// </summary>
+        /// <param name="lines">Line segments of a polygon that represents the gradient area.</param>
+        /// <param name="colors">Array of colors that correspond to each point in the polygon.</param>
+        public PathGradientBrush(ILineSegment[] lines, Color[] colors)
+            : this(lines, colors, CalculateCenterColor(colors))
+        {
+        }
+
+        /// <inheritdoc />
+        public BrushApplicator<TPixel> CreateApplicator<TPixel>(
+            ImageFrame<TPixel> source,
+            RectangleF region,
+            GraphicsOptions options)
+            where TPixel : struct, IPixel<TPixel>
+        {
+            return new PathGradientBrushApplicator<TPixel>(source, this.path, this.edges, this.centerColor, options);
+        }
+
+        private static Color CalculateCenterColor(Color[] colors) =>
+            new Color(colors.Select(c => c.ToVector4()).Aggregate((p1, p2) => p1 + p2) / colors.Length);
+
+        private static float DistanceBetween(PointF p1, PointF p2) => ((Vector2)(p2 - p1)).Length();
+
+        /// <summary>
+        /// An edge of the polygon that represents the gradient area.
+        /// </summary>
+        private class Edge
+        {
+            private readonly Path path;
+
+            private readonly float length;
+
+            private readonly PointF[] buffer;
+
+            public Edge(Path path, Color startColor, Color endColor)
+            {
+                this.path = path;
+
+                Vector2[] points = path.LineSegments.SelectMany(s => s.Flatten()).Select(p => (Vector2)p).ToArray();
+
+                this.Start = points.First();
+                this.StartColor = startColor.ToVector4();
+
+                this.End = points.Last();
+                this.EndColor = endColor.ToVector4();
+
+                this.length = DistanceBetween(this.End, this.Start);
+                this.buffer = new PointF[this.path.MaxIntersections];
+            }
+
+            public PointF Start { get; }
+
+            public Vector4 StartColor { get; }
+
+            public PointF End { get; }
+
+            public Vector4 EndColor { get; }
+
+            public (PointF point, float distance)? FindIntersection(PointF start, PointF end)
+            {
+                int intersections = this.path.FindIntersections(start, end, this.buffer);
+
+                if (intersections == 0)
+                {
+                    return null;
+                }
+
+                return this.buffer
+                    .Take(intersections)
+                    .Select(p => (point: p, distance: ((Vector2)(p - start)).Length()))
+                    .OrderBy(v => v.distance)
```

Comment:
Another optimization hint:
Finding a closest intersection might be enough you don't need to order the whole ordered list.

Reply:
Removed unnecessary `OrderBy` invocations. Thanks!